### DISCUSSION TITLE
How many routes to handle get One vs All vs Some?

### DISCUSSION THREAD

**Discussion by paulwongx** [trpc/trpc#3562] [Q&A] [answered]
I have a typical route that pulls data, let's say it's `posts` for example. Currently it's one route and you can pass in no arguments to get all, or one argument to get one. However, Typescript isn't correctly returning the proper type based on the arguments passed in. It thinks the route returns `IPost` and `IPost[]`, so when I use the data on the front end I have to keep casting the return type. 

Is it best practice to have multiple routes for this instead? One for getOnePost, getAllPosts, getSomePosts? Or is it possible for Typescript to understand conditional return types in a function? 

Currently my code looks something like:
```js
publicProcedure
    .input(z.object({ id: z.string().nullish() }).nullish())
    .query(async ({ input }) => {
      const url = "https://example.com/posts";

      const json = (await got.get(url).json()) as IPosts[];
      if (input && input.id) {
        const posts = json.find((p) => p.id=== input.id);
        if (posts) return posts;
      }

      return json;
    }),
```
Based on this StackOverflow answer, I see you can conditionally cast the function. How would that look like for an unnamed function?
https://stackoverflow.com/a/55059896/12894626

**Comment by KATT** [comment, marked as answer]
Conditional return types are not supported, see #2150
